File: runtime/shims.py
"""
shims.py -- host implementations of the 99 imports (milestone M2)

Every function libHAL.Android.so can reach outside itself lands here.  The
list is closed: the library has no dlopen and issues no syscalls, so nothing
can appear at runtime that is not registered below.
"""
import math, os, struct, time, errno as _errno
from guest import Guest, GuestError

REG = {}
def shim(*names):
    def deco(fn):
        for n in names:
            REG[n] = fn
        return fn
    return deco


def install(g, root=None, log=False):
    g.log_shims = log
    g.root = root or os.path.abspath("guestfs")
    os.makedirs(g.root, exist_ok=True)
    g.files = {}                 # guest FILE* -> python handle
    g.gl_log = []                # recorded GL calls (no GL on the desktop harness)
    for n, fn in REG.items():
        g.shims[n] = fn
    for n in g.imports:
        if n.startswith('gl') and n not in g.shims:
            g.shims[n] = _gl_record(n)
    return g


# ------------------------------------------------------------------ memory
@shim('malloc')
def _malloc(g):    g.ret(g.malloc(g.arg(0)))

@shim('free')
def _free(g):      g.free(g.arg(0)); g.ret(0)

@shim('realloc')
def _realloc(g):
    p, n = g.arg(0), g.arg(1)
    if not p:
        return g.ret(g.malloc(n))
    old = g.blocks.get(p, [0, False])[0]
    new = g.malloc(n)
    if old:
        g.write(new, g.read(p, min(old, n)))
    g.free(p)
    g.ret(new)

@shim('memcpy', 'memmove')
def _memcpy(g):
    d, s, n = g.arg(0), g.arg(1), g.arg(2)
    if n:
        g.write(d, g.read(s, n))
    g.ret(d)

@shim('memset')
def _memset(g):
    d, c, n = g.arg(0), g.arg(1) & 0xff, g.arg(2)
    if n:
        g.write(d, bytes([c]) * n)
    g.ret(d)


# -------------------------------------------------------------------- math
def _d1(fn):
    def f(g): g.retd(fn(g.argd(0)))
    return f

def _d2(fn):
    def f(g): g.retd(fn(g.argd(0), g.argd(2)))
    return f

def _safe(fn, default=0.0):
    def w(*a):
        try:    return fn(*a)
        except (ValueError, OverflowError, ZeroDivisionError): return default
    return w

for _n, _f in [('sin', math.sin), ('cos', math.cos), ('tan', math.tan),
               ('asin', math.asin), ('acos', math.acos), ('atan', math.atan),
               ('sqrt', math.sqrt), ('log', math.log)]:
    REG[_n] = _d1(_safe(_f))
REG['atan2'] = _d2(_safe(math.atan2))
REG['pow']   = _d2(_safe(math.pow))


# -------------------------------------------------------------------- time
@shim('clock_gettime')
def _clock_gettime(g):
    ts = g.arg(1)
    t = time.perf_counter()
    g.write(ts, struct.pack('<II', int(t), int((t % 1) * 1e9)))
    g.ret(0)

@shim('time')
def _time(g):
    t = int(time.time())
    if g.arg(0):
        g.write(g.arg(0), struct.pack('<I', t))
    g.ret(t)

@shim('localtime')
def _localtime(g):
    tp = g.arg(0)
    secs = struct.unpack('<I', g.read(tp, 4))[0] if tp else int(time.time())
    lt = time.localtime(secs)
    buf = getattr(g, '_tm', None) or g.alloc_scratch(44)
    g._tm = buf
    g.write(buf, struct.pack('<9i', lt.tm_sec, lt.tm_min, lt.tm_hour, lt.tm_mday,
                             lt.tm_mon - 1, lt.tm_year - 1900, lt.tm_wday,
                             lt.tm_yday, lt.tm_isdst))
    g.ret(buf)

@shim('usleep')
def _usleep(g):
    # real sleeping would stall the harness; the engine only uses this to yield
    g.ret(0)


# ---------------------------------------------------------------- file I/O
def _hostpath(g, p):
    p = (p or '').replace('\\', '/').lstrip('/')
    return os.path.join(g.root, p)

@shim('fopen')
def _fopen(g):
    path, mode = g.cstr(g.arg(0)), (g.cstr(g.arg(1)) or 'rb')
    m = mode.replace('b', '') + 'b'
    try:
        fh = open(_hostpath(g, path), m)
    except OSError:
        return g.ret(0)
    fp = g.malloc(8)
    g.files[fp] = fh
    if getattr(g, 'log_shims', False):
        print("      fopen(%s,%s) -> 0x%x" % (path, mode, fp))
    g.ret(fp)

@shim('fdopen')
def _fdopen(g):
    """The engine receives its asset pack as a file descriptor from
    Java (openRawResourceFd) and wraps it here."""
    fd, mode = g.arg(0), g.cstr(g.arg(1)) or 'rb'
    fh = getattr(g, 'fd_files', {}).get(fd)
    if fh is None:
        return g.ret(0)
    fp = g.malloc(8)
    g.files[fp] = fh
    if getattr(g, 'log_shims', False):
        print("      fdopen(fd=%d,%s) -> FILE 0x%x" % (fd, mode, fp))
    g.ret(fp)

@shim('fread')
def _fread(g):
    buf, sz, n, fp = g.arg(0), g.arg(1), g.arg(2), g.arg(3)
    fh = g.files.get(fp)
    if not fh or sz * n == 0:
        return g.ret(0)
    data = fh.read(sz * n)
    g.write(buf, data)
    g.ret(len(data) // sz if sz else 0)

@shim('fwrite')
def _fwrite(g):
    buf, sz, n, fp = g.arg(0), g.arg(1), g.arg(2), g.arg(3)
    fh = g.files.get(fp)
    if not fh or sz * n == 0:
        return g.ret(0)
    fh.write(g.read(buf, sz * n))
    g.ret(n)

@shim('fseek')
def _fseek(g):
    fh = g.files.get(g.arg(0))
    if not fh:
        return g.ret(-1 & 0xffffffff)
    off = g.arg(1)
    fh.seek(off - (1 << 32) if off >= (1 << 31) else off, g.arg(2))
    g.ret(0)

@shim('ftell')
def _ftell(g):
    fh = g.files.get(g.arg(0))
    g.ret(fh.tell() if fh else -1 & 0xffffffff)

@shim('fclose')
def _fclose(g):
    fh = g.files.pop(g.arg(0), None)
    if fh:
        fh.close()
    g.ret(0)

@shim('fflush')
def _fflush(g):
    fh = g.files.get(g.arg(0))
    if fh:
        fh.flush()
    g.ret(0)

@shim('fputc')
def _fputc(g):  g.ret(g.arg(0))

@shim('stat')
def _stat(g):
    try:
        os.stat(_hostpath(g, g.cstr(g.arg(0))))
        g.ret(0)
    except OSError:
        g.ret(-1 & 0xffffffff)

@shim('mkdir')
def _mkdir(g):
    try:
        os.makedirs(_hostpath(g, g.cstr(g.arg(0))), exist_ok=True)
        g.ret(0)
    except OSError:
        g.ret(-1 & 0xffffffff)

@shim('remove')
def _remove(g):
    try:
        os.remove(_hostpath(g, g.cstr(g.arg(0))))
        g.ret(0)
    except OSError:
        g.ret(-1 & 0xffffffff)

@shim('chdir')
def _chdir(g):  g.ret(0)

@shim('dup')
def _dup(g):    g.ret(g.arg(0))          # same descriptor is fine here

@shim('close', 'fcntl')
def _fdops(g):  g.ret(0)


# ------------------------------------------------------------- varargs I/O
def _fmt(g, fmt, argi):
    """Render a C format string using ARM32 varargs starting at arg index."""
    out, i, n = [], 0, argi
    while i < len(fmt):
        c = fmt[i]
        if c != '%':
            out.append(c); i += 1; continue
        j = i + 1
        while j < len(fmt) and fmt[j] in '-+ #0123456789.*lhqLzjt':
            j += 1
        if j >= len(fmt):
            break
        conv, spec = fmt[j], fmt[i:j + 1]
        longlong = 'll' in spec or 'q' in spec
        spec = spec.replace('ll', '').replace('l', '').replace('h', '').replace('q', '')
        try:
            if conv == '%':
                out.append('%')
            elif conv in 'di':
                if longlong:
                    if n % 2: n += 1
                    v = g.arg(n) | (g.arg(n + 1) << 32); n += 2
                else:
                    v = g.arg(n); n += 1
                    v = v - (1 << 32) if v >= (1 << 31) else v
                out.append(spec % v)
            elif conv in 'uxXo':
                v = g.arg(n); n += 1
                out.append(spec % v)
            elif conv in 'fFeEgG':
                if n % 2: n += 1
                out.append(spec % g.argd(n)); n += 2
            elif conv == 'c':
                out.append(chr(g.arg(n) & 0xff)); n += 1
            elif conv == 's':
                out.append(g.cstr(g.arg(n)) or "(null)"); n += 1
            elif conv == 'p':
                out.append("0x%x" % g.arg(n)); n += 1
            else:
                out.append(spec)
        except Exception:
            out.append(spec)
        i = j + 1
    return ''.join(out)

@shim('__android_log_print')
def _logprint(g):
    tag, fmt = g.cstr(g.arg(1)), g.cstr(g.arg(2)) or ''
    msg = _fmt(g, fmt, 3)
    print("      [logcat/%s] %s" % (tag, msg.rstrip()))
    g.ret(0)

@shim('fprintf')
def _fprintf(g):
    fmt = g.cstr(g.arg(1)) or ''
    msg = _fmt(g, fmt, 2)
    fh = g.files.get(g.arg(0))
    if fh:
        fh.write(msg.encode('utf-8', 'replace'))
    else:
        print("      [stderr] %s" % msg.rstrip())
    g.ret(len(msg))


# ----------------------------------------------------------------- runtime
@shim('__errno')
def _errno_(g):  g.ret(g.errno_addr)

@shim('abort')
def _abort(g):   raise GuestError("guest called abort()")

@shim('__stack_chk_fail')
def _sschk(g):   raise GuestError("stack smashing detected in guest")

@shim('__gnu_Unwind_Find_exidx')
def _find_exidx(g):
    base, count = g.exidx
    if g.arg(1):
        g.write(g.arg(1), struct.pack('<I', count))
    g.ret(base)

@shim('__cxa_begin_cleanup', '__cxa_type_match', '__cxa_call_unexpected')
def _cxa(g):     g.ret(1)


# ----------------------------------------------------------------- sockets
# Failure conventions differ and it matters: the syscall-style calls report
# -1, but gethostbyname returns a POINTER and must report NULL.  Returning
# -1 there makes the engine treat 0xffffffff as a valid struct hostent* and
# dereference it -- which is exactly how UE_ResolveHostAddress crashed.
def _fail_minus1(g):  g.ret(0xffffffff)
def _fail_null(g):    g.ret(0)

for _n in ('socket', 'bind', 'connect', 'send', 'recv', 'sendto', 'recvfrom',
           'sendmsg', 'setsockopt', 'gethostname'):
    REG[_n] = _fail_minus1
REG['gethostbyname'] = _fail_null


# ---------------------------------------------------------------- graphics
def _gl_record(name):
    """Desktop harness has no GL context; record the call and continue.
    On device these become direct forwards to the real driver."""
    def f(g):
        g.gl_log.append((name, [g.arg(i) for i in range(10)]))
        if name in ('glGenTextures', 'glGenBuffers'):
            cnt, ptr = g.arg(0), g.arg(1)
            for i in range(cnt):
                g._glid = getattr(g, '_glid', 0) + 1
                g.write(ptr + i * 4, struct.pack('<I', g._glid))
        g.ret(0)
    return f

File: android/app/src/main/cpp/shims.cpp
// shims.cpp -- host implementations of the 99 imports.
//
// Ported from runtime/shims.py. The import set is closed: the library has no
// dlopen and issues no SVC, so nothing can appear here at runtime that is not
// registered below.
//
// One convention is deliberately non-uniform and must stay that way:
// gethostbyname returns a POINTER and reports failure as NULL. Stubbing it to
// -1 alongside the socket calls made the engine accept 0xffffffff as a valid
// struct hostent* and dereference it -- a crash that only appeared 169 frames
// into a running game, inside UE_ResolveHostAddress.
#include "shogun_runtime.h"

#include <android/log.h>
#include <sys/stat.h>
#include <unistd.h>

#include <cmath>
#include <cstdarg>
#include <cstdio>
#include <cstring>
#include <ctime>
#include <string>
#include <vector>
#include <unordered_map>

namespace shogun {
namespace {

constexpr const char* kTag = "shogun";

struct FileState {
  std::string root;                                  // app-private storage
  std::unordered_map<uint32_t, FILE*> open;          // guest FILE* -> host
  std::unordered_map<int, FILE*> fds;                // fd -> host (asset pack)
  uint32_t tm_buf = 0;
};
FileState* g_files = nullptr;

std::string HostPath(const std::string& p) {
  std::string s = p;
  for (auto& c : s) if (c == '\\') c = '/';
  // The engine builds a path by pasting the storage directory it was given
  // onto a file name, so what arrives here is already absolute and already
  // inside our sandbox. Stripping the leading slash and re-prefixing turned
  // it into <root>/<root>/<name>, and every single save failed to open.
  if (s.size() > g_files->root.size() &&
      s.compare(0, g_files->root.size(), g_files->root) == 0)
    return s;
  while (!s.empty() && s.front() == '/') s.erase(s.begin());
  return g_files->root + "/" + s;
}

// ---- ARM32 varargs -------------------------------------------------------
// Walk r1-r3 then the stack, using the format string to know each width.
std::string FormatVa(Runtime& rt, const std::string& fmt, int argi) {
  std::string out;
  size_t i = 0;
  int n = argi;
  char tmp[512];
  while (i < fmt.size()) {
    if (fmt[i] != '%') { out.push_back(fmt[i++]); continue; }
    size_t j = i + 1;
    while (j < fmt.size() && std::strchr("-+ #0123456789.*lhqLzjt", fmt[j])) j++;
    if (j >= fmt.size()) break;
    char conv = fmt[j];
    std::string spec = fmt.substr(i, j - i + 1);
    bool ll = spec.find("ll") != std::string::npos || spec.find('q') != std::string::npos;
    // strip length modifiers; we supply the right C type ourselves
    std::string clean;
    for (size_t k = 0; k < spec.size(); k++)
      if (!std::strchr("lhqLzjt", spec[k])) clean.push_back(spec[k]);
    tmp[0] = 0;
    if (conv == '%') { out.push_back('%'); i = j + 1; continue; }
    if (conv == 'd' || conv == 'i') {
      if (ll) {
        if (n % 2) n++;
        int64_t v = static_cast<int64_t>(
            static_cast<uint64_t>(rt.Arg(n)) | (static_cast<uint64_t>(rt.Arg(n + 1)) << 32));
        n += 2;
        std::snprintf(tmp, sizeof tmp, (clean.substr(0, clean.size() - 1) + "lld").c_str(), (long long)v);
      } else {
        std::snprintf(tmp, sizeof tmp, clean.c_str(), static_cast<int32_t>(rt.Arg(n++)));
      }
    } else if (std::strchr("uxXo", conv)) {
      std::snprintf(tmp, sizeof tmp, clean.c_str(), rt.Arg(n++));
    } else if (std::strchr("fFeEgG", conv)) {
      if (n % 2) n++;
      double d = rt.ArgD(n); n += 2;
      std::snprintf(tmp, sizeof tmp, clean.c_str(), d);
    } else if (conv == 'c') {
      std::snprintf(tmp, sizeof tmp, clean.c_str(), static_cast<int>(rt.Arg(n++) & 0xff));
    } else if (conv == 's') {
      std::string s = rt.CStr(rt.Arg(n++));
      std::snprintf(tmp, sizeof tmp, clean.c_str(), s.c_str());
    } else if (conv == 'p') {
      std::snprintf(tmp, sizeof tmp, "0x%x", rt.Arg(n++));
    } else {
      std::snprintf(tmp, sizeof tmp, "%s", clean.c_str());
    }
    out += tmp;
    i = j + 1;
  }
  return out;
}

}  // namespace

void InstallShims(Runtime& rt, const std::string& files_dir) {
  static FileState state;
  state.root = files_dir;
  g_files = &state;

  auto S = [&rt](const char* n, ShimFn f) { rt.SetShim(n, std::move(f)); };

  // ------------------------------------------------------------- memory
  S("malloc",  [](Runtime& r) { r.Ret(r.Malloc(r.Arg(0))); });
  S("free",    [](Runtime& r) { r.Free(r.Arg(0)); r.Ret(0); });
  S("realloc", [](Runtime& r) {
    uint32_t p = r.Arg(0), n = r.Arg(1);
    if (!p) { r.Ret(r.Malloc(n)); return; }
    uint32_t np = r.Malloc(n);
    if (np && n) {
      std::vector<uint8_t> buf(n);
      r.Read(p, buf.data(), n);          // copies at most the new size
      r.Write(np, buf.data(), n);
    }
    r.Free(p);
    r.Ret(np);
  });
  auto copy = [](Runtime& r) {
    uint32_t d = r.Arg(0), s = r.Arg(1), n = r.Arg(2);
    if (n) {
      std::vector<uint8_t> buf(n);
      r.Read(s, buf.data(), n);
      r.Write(d, buf.data(), n);
    }
    r.Ret(d);
  };
  S("memcpy", copy);
  S("memmove", copy);
  S("memset", [](Runtime& r) {
    uint32_t d = r.Arg(0), n = r.Arg(2);
    uint8_t c = static_cast<uint8_t>(r.Arg(1));
    if (n) { std::vector<uint8_t> buf(n, c); r.Write(d, buf.data(), n); }
    r.Ret(d);
  });

  // --------------------------------------------------------------- math
  // Soft-float: doubles arrive in an even register pair and return in r0:r1.
  auto d1 = [](double (*fn)(double)) {
    return [fn](Runtime& r) { r.RetD(fn(r.ArgD(0))); };
  };
  auto d2 = [](double (*fn)(double, double)) {
    return [fn](Runtime& r) { r.RetD(fn(r.ArgD(0), r.ArgD(2))); };
  };
  S("sin",  d1(std::sin));   S("cos",  d1(std::cos));   S("tan", d1(std::tan));
  S("asin", d1(std::asin));  S("acos", d1(std::acos));  S("atan", d1(std::atan));
  S("sqrt", d1(std::sqrt));  S("log",  d1(std::log));
  S("atan2", d2(std::atan2)); S("pow", d2(std::pow));

  // --------------------------------------------------------------- time
  S("clock_gettime", [](Runtime& r) {
    struct timespec ts{};
    clock_gettime(CLOCK_MONOTONIC, &ts);
    uint32_t v[2] = {static_cast<uint32_t>(ts.tv_sec),
                     static_cast<uint32_t>(ts.tv_nsec)};
    r.Write(r.Arg(1), v, sizeof v);
    r.Ret(0);
  });
  S("time", [](Runtime& r) {
    uint32_t t = static_cast<uint32_t>(::time(nullptr));
    if (r.Arg(0)) r.Write(r.Arg(0), &t, 4);
    r.Ret(t);
  });
  S("localtime", [](Runtime& r) {
    uint32_t tp = r.Arg(0), secs = 0;
    if (tp) r.Read(tp, &secs, 4); else secs = static_cast<uint32_t>(::time(nullptr));
    time_t t = secs;
    struct tm lt{};
    localtime_r(&t, &lt);
    if (!g_files->tm_buf) g_files->tm_buf = r.AllocScratch(44);
    int32_t v[9] = {lt.tm_sec, lt.tm_min, lt.tm_hour, lt.tm_mday,
                    lt.tm_mon, lt.tm_year, lt.tm_wday, lt.tm_yday, lt.tm_isdst};
    r.Write(g_files->tm_buf, v, sizeof v);
    r.Ret(g_files->tm_buf);
  });
  // The engine only uses usleep to yield; real sleeping would stall the frame.
  S("usleep", [](Runtime& r) { r.Ret(0); });

  // ------------------------------------------------------------ file I/O
  S("fopen", [](Runtime& r) {
    std::string path = r.CStr(r.Arg(0));
    std::string mode = r.CStr(r.Arg(1));
    if (mode.empty()) mode = "rb";
    if (mode.find('b') == std::string::npos) mode += "b";
    FILE* f = std::fopen(HostPath(path).c_str(), mode.c_str());
    // Saving is the one thing that has to be visible from outside the process,
    // and the app is not debuggable, so every write-mode open gets a line.
    if (!f) {
      __android_log_print(f ? ANDROID_LOG_INFO : ANDROID_LOG_WARN, kTag,
          "fopen(\"%s\", \"%s\") -> %s   [%s]", path.c_str(), mode.c_str(),
          f ? "ok" : "FAILED", HostPath(path).c_str());
    }
    if (!f) { r.Ret(0); return; }
    uint32_t fp = r.Malloc(8);
    g_files->open[fp] = f;
    r.Ret(fp);
  });
  // The asset pack arrives as a descriptor from Java (openRawResourceFd).
  S("fdopen", [](Runtime& r) {
    int fd = static_cast<int>(r.Arg(0));
    auto it = g_files->fds.find(fd);
    if (it == g_files->fds.end()) { r.Ret(0); return; }
    uint32_t fp = r.Malloc(8);
    g_files->open[fp] = it->second;
    r.Ret(fp);
  });
  S("fread", [](Runtime& r) {
    uint32_t buf = r.Arg(0), sz = r.Arg(1), n = r.Arg(2), fp = r.Arg(3);
    auto it = g_files->open.find(fp);
    if (it == g_files->open.end() || !sz || !n) { r.Ret(0); return; }
    std::vector<uint8_t> tmp(static_cast<size_t>(sz) * n);
    size_t got = std::fread(tmp.data(), 1, tmp.size(), it->second);
    if (got) r.Write(buf, tmp.data(), got);
    r.Ret(static_cast<uint32_t>(got / sz));
  });
  S("fwrite", [](Runtime& r) {
    uint32_t buf = r.Arg(0), sz = r.Arg(1), n = r.Arg(2), fp = r.Arg(3);
    auto it = g_files->open.find(fp);
    if (it == g_files->open.end() || !sz || !n) { r.Ret(0); return; }
    std::vector<uint8_t> tmp(static_cast<size_t>(sz) * n);
    r.Read(buf, tmp.data(), tmp.size());
    std::fwrite(tmp.data(), 1, tmp.size(), it->second);
    r.Ret(n);
  });
  S("fseek", [](Runtime& r) {
    auto it = g_files->open.find(r.Arg(0));
    if (it == g_files->open.end()) { r.Ret(0xFFFFFFFFu); return; }
    std::fseek(it->second, static_cast<int32_t>(r.Arg(1)), static_cast<int>(r.Arg(2)));
    r.Ret(0);
  });
  S("ftell", [](Runtime& r) {
    auto it = g_files->open.find(r.Arg(0));
    r.Ret(it == g_files->open.end() ? 0xFFFFFFFFu
                                    : static_cast<uint32_t>(std::ftell(it->second)));
  });
  S("fclose", [](Runtime& r) {
    auto it = g_files->open.find(r.Arg(0));
    if (it != g_files->open.end()) { std::fclose(it->second); g_files->open.erase(it); }
    r.Ret(0);
  });
  S("fflush", [](Runtime& r) {
    auto it = g_files->open.find(r.Arg(0));
    if (it != g_files->open.end()) std::fflush(it->second);
    r.Ret(0);
  });
  S("fputc", [](Runtime& r) { r.Ret(r.Arg(0)); });
  S("stat", [](Runtime& r) {
    FILE* f = std::fopen(HostPath(r.CStr(r.Arg(0))).c_str(), "rb");
    if (f) { std::fclose(f); r.Ret(0); } else r.Ret(0xFFFFFFFFu);
  });
  S("mkdir", [](Runtime& r) {
    std::string p = HostPath(r.CStr(r.Arg(0)));
    std::string cmd;
    // create the whole chain; the engine makes nested save directories
    for (size_t i = 1; i <= p.size(); i++) {
      if (i == p.size() || p[i] == '/') {
        std::string part = p.substr(0, i);
        ::mkdir(part.c_str(), 0770);
      }
    }
    r.Ret(0);
  });
  S("remove", [](Runtime& r) {
    r.Ret(std::remove(HostPath(r.CStr(r.Arg(0))).c_str()) == 0 ? 0 : 0xFFFFFFFFu);
  });
  S("chdir", [](Runtime& r) { r.Ret(0); });
  S("dup",   [](Runtime& r) { r.Ret(r.Arg(0)); });   // same descriptor is fine
  S("close", [](Runtime& r) { r.Ret(0); });
  S("fcntl", [](Runtime& r) { r.Ret(0); });

  // ------------------------------------------------------- varargs output
  S("__android_log_print", [](Runtime& r) {
    std::string tag = r.CStr(r.Arg(1));
    std::string fmt = r.CStr(r.Arg(2));
    std::string msg = FormatVa(r, fmt, 3);
    __android_log_print(ANDROID_LOG_INFO, tag.empty() ? kTag : tag.c_str(),
                        "%s", msg.c_str());
    r.Ret(0);
  });
  S("fprintf", [](Runtime& r) {
    std::string fmt = r.CStr(r.Arg(1));
    std::string msg = FormatVa(r, fmt, 2);
    auto it = g_files->open.find(r.Arg(0));
    if (it != g_files->open.end()) std::fwrite(msg.data(), 1, msg.size(), it->second);
    else __android_log_print(ANDROID_LOG_WARN, kTag, "%s", msg.c_str());
    r.Ret(static_cast<uint32_t>(msg.size()));
  });

  // ------------------------------------------------------------- runtime
  S("__errno", [](Runtime& r) { r.Ret(r.ErrnoAddr()); });
  S("abort", [](Runtime& r) {
    __android_log_print(ANDROID_LOG_ERROR, kTag, "guest called abort()");
    r.Ret(0);
  });
  S("__stack_chk_fail", [](Runtime& r) {
    __android_log_print(ANDROID_LOG_ERROR, kTag, "guest stack smashing detected");
    r.Ret(0);
  });
  // The unwinder itself lives inside the library and runs emulated; we only
  // have to hand it the .ARM.exidx bounds.
  S("__gnu_Unwind_Find_exidx", [](Runtime& r) {
    uint32_t count = r.ExidxCount();
    if (r.Arg(1)) r.Write(r.Arg(1), &count, 4);
    r.Ret(r.ExidxBase());
  });
  for (const char* n : {"__cxa_begin_cleanup", "__cxa_type_match",
                        "__cxa_call_unexpected"})
    S(n, [](Runtime& r) { r.Ret(1); });

  // ------------------------------------------------------------- sockets
  // Vestigial (adhocwifi.c / gamecenter.c). Note the split convention.
  for (const char* n : {"socket", "bind", "connect", "send", "recv", "sendto",
                        "recvfrom", "sendmsg", "setsockopt", "gethostname"})
    S(n, [](Runtime& r) { r.Ret(0xFFFFFFFFu); });
  S("gethostbyname", [](Runtime& r) { r.Ret(0); });   // returns a pointer: NULL
}

// Lets the JNI layer hand the asset-pack descriptor to fdopen.
void RegisterAssetFd(int fd, FILE* f) {
  if (g_files) g_files->fds[fd] = f;
}

}  // namespace shogun
